lib/pbench/cli/server/shell.py

```diff
@@ -37,6 +42,57 @@ def find_the_unicorn(logger: Logger):
             )
 
 
+def keycloak_connection(config: PbenchServerConfig, logger: Logger):
+    """
+     Checks if the Keycloak server is up and accepting the connections.
+     The connection check does the GET request on the oidc server /health
+     endpoint and the sample response returned by the /health endpoint looks
+     like the following:
+        {
+            "status": "UP", # if the server is up
+            "checks": []
+        }
+     Note: The Keycloak server need to be started with health-enabled on.
+     Args:
+        config: PbenchServerConfig
+        logger: logger
+    Returns:
+        0 if successful
+    """
+    try:
+        oidc_server = config.get(
+            "authentication",
+            "internal_server_url",
+            fallback=config.get("authentication", "server_url"),
+        )
+    except (NoSectionError):
+        logger.exception("Bad config file:  missing 'authentication' section")
+        sys.exit(1)
+    except (NoOptionError):
+        logger.error("Bad config file:  no 'server_url' in  'authentication' section")
+        sys.exit(1)
+
+    session = requests.Session()
+    # The connection check will retry thrice unless successful, viz.,
+    # [0.0s, 20.0s, 40.0s]. urllib3 will sleep for:
+    # {backoff factor} * (2 ^ ({number of total retries} - 1)) seconds between
+    # the retries. More detailed documentation on Retry and backoff_factor can
+    # be found at:
+    # https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry
+    retry = Retry(total=3, backoff_factor=10)
```

Comment:
Given that the principal use of this is to cover bootstrapping sequencing, it strikes me that only three retries, even with a factor of 10, doesn't cover enough time (i.e., the first one is immediate; the second happens after 10 seconds, and the third after 20 -- so that's an overall delay of much less than a minute -- and we should probably be covering at least two, perhaps five, minutes).

I don't think the individual requests cost very much, so we can afford to do much more than three, and, given that the backoff is exponential, I think we can afford a much smaller backoff factor.  So, I would recommend a factor between 0.5 and 2 with a `total` of 8 (which will give us a timeout between 2 and 8.5 minutes, I think).

Reply:
With the factor of 10 _I think_ the retries will happen at ` [0.0s, 20.0s, 40.0s]` so that's 3 tries in a minute. It is calculated as 
```{backoff factor} * (2 ^ ({number of total retries} - 1)) seconds between the retries after the first attempt```

so 
1. 0.0s 
2. 10 * (2 ^ (2-1)) = 10 * 2 = 20.0s #2nd retry 
3. 10 * (2 ^ (3-1)) = 10 * 2 = 20.0s #3rd retry

Reply:
You're quite right (although, I think you have a couple of typos in the last line, there 😉), your configuration covers a full minute; however, I still think that that is too short.

So, I think you should increase the `total` number...and, if you're going to do that, I would reduce the backoff factor as well (just so long as the net result covers at least two minutes and perhaps as much as five).

Reply:
yes it should be 10 * (2 ^ (3-1)) = 10 * 4 = 40.0s #3rd retry :slightly_smiling_face:

Reply:
(You can just edit [your reply](https://github.com/distributed-system-analysis/pbench/pull/3065#discussion_r1029781385) and fix it, y'know!  😁 )

Reply:
But then if someone else reads the conversation then they will be confused why Webb said there is a typo when there is no typo :rofl:

Reply:
Yes, but, once there is no longer any typo, I'll remove the comment from my previous reply and then delete the subsequent one(s).  😁